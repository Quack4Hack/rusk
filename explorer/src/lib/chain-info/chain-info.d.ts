type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "block" | "transaction";
};

type Transaction = {
  blockhash: string;
  blockheight: number;
  contract: string;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
};
