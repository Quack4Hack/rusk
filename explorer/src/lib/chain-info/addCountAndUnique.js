/**
 * Returns an array of unique objects, sorted by the frequency of their occurrence,
 * with an additional count property representing the number of times
 * that city-country combination appeared in the input data.
 *
 * @param {Omit<NodeLocation, "count">[]} data
 * @returns {NodeLocation[]}
 */
function addCountAndUnique(data) {
  const countMap = data.reduce((locationMap, location) => {
    const key = `${location.city},${location.country}`;

    if (!locationMap[key]) {
      locationMap[key] = { ...location, count: 0 };
    }
    locationMap[key].count += 1;
    return locationMap;
  }, /** @type {{ [key: string]: NodeLocation }} */ ({}));

  // Return only unique city-country combinations
  return Object.values(countMap).sort((a, b) => b.count - a.count);
}

export default addCountAndUnique;
