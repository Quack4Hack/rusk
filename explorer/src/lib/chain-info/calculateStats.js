import {
  always,
  compose,
  condition,
  filterWith,
  getKey,
  partition,
} from "lamb";

import { arraySumByKey } from "$lib/dusk/array";

/** @type {(txs: Pick<GQLTransaction, "err">[]) => number} */
const getFailedTxAmount = compose(
  getKey("length"),
  filterWith((tx) => tx.err !== null)
);

/**
 * We take into account only provisioners with
 * the minimum stake amount of 1000 Dusk (1e12 Lux).
 *
 * @type {(provisioners: HostProvisioner[]) => HostProvisioner[]}
 */
const getValidProvisioners = filterWith((p) => p.amount >= 1e12);

/**
 * Sums the values of the "amount" key in the received array.
 * Returns zero if the array is empty.
 */
const sumByAmount = condition(
  (provisioners) => provisioners.length > 0,
  arraySumByKey("amount"),
  always(0)
);

/**
 * @param {HostProvisioner[]} provisioners
 * @param {number} lastHeight
 * @param {Pick<GQLTransaction, "err">[]} last100BlocksTxs
 * @returns {Stats}
 */
function calculateStats(provisioners, lastHeight, last100BlocksTxs) {
  const [activeProvisioners, waitingProvisioners] = partition(
    getValidProvisioners(provisioners),
    (p) => p.eligibility <= lastHeight
  );

  return {
    activeProvisioners: activeProvisioners.length,
    activeStake: sumByAmount(activeProvisioners),
    lastBlock: lastHeight,
    txs100blocks: {
      failed: getFailedTxAmount(last100BlocksTxs),
      transfers: last100BlocksTxs.length,
    },
    waitingProvisioners: waitingProvisioners.length,
    waitingStake: sumByAmount(waitingProvisioners),
  };
}

export default calculateStats;
