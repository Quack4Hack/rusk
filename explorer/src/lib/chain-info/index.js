export { default as transformBlock } from "./transformBlock";
export { default as transformTransaction } from "./transformTransaction";
