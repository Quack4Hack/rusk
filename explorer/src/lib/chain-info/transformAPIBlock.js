import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformAPITransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformAPIBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts", "version"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformAPIBlock = ({ header, transactions }) => ({
  header: transformAPIBlockHeader(header),
  transactions: updateIn(
    transactions,
    "data",
    mapWith(transformAPITransaction)
  ),
});

export default transformAPIBlock;
