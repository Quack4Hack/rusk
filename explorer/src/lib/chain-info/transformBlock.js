import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformTransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformBlock = ({ header, transactions }) => ({
  header: transformBlockHeader(header),
  transactions: updateIn(transactions, "data", mapWith(transformTransaction)),
});

export default transformBlock;
