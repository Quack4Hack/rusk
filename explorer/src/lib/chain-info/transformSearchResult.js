import { compose, getPathIn, mapWith } from "lamb";

/**
 * @template {Record<PropertyKey, any>} T
 * @template {string} U
 * @param {T} source
 * @param {U} path
 * @returns {Exclude<import("lamb").GetPath<T, U>, undefined> | []}
 */
const getPathOrEmptyArray = (source, path) => getPathIn(source, path) ?? [];

/** @type {(v: APISearchResult) => SearchResult[]} */
const transformSearchResult = compose(
  mapWith((entry) =>
    entry?.header?.hash
      ? { id: entry.header.hash, type: "block" }
      : { id: entry.txid, type: "transaction" }
  ),
  (v) => [
    ...getPathOrEmptyArray(v, "data.data.blocks"),
    ...getPathOrEmptyArray(v, "data.data.transactions"),
  ]
);

export default transformSearchResult;
