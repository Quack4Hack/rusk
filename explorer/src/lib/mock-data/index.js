export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactionDetails } from "./gql-transaction-details.json";
export { default as gqlTransactions } from "./gql-transactions.json";
