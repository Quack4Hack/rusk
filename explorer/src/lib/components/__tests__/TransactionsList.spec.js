import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";

import { gqlTransaction } from "$lib/mock-data";
import { transformTransaction } from "$lib/chain-info";

import { TransactionsList } from "..";

/** @param {HTMLElement} container */
function getTimeElement(container) {
  return /** @type {HTMLTimeElement} */ (container.querySelector("time"));
}

describe("Transactions List", () => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(2024, 4, 20));

  const transaction = transformTransaction(gqlTransaction.tx);

  /** @type {import("svelte").ComponentProps<TransactionsList>} */
  const baseProps = {
    data: transaction,
    mode: "full",
  };

  const timeRefreshInterval = 1000;

  afterEach(cleanup);

  afterAll(() => {
    vi.useRealTimers();
  });

  it('should render the `TransactionsList` component in "full" mode', () => {
    const { container } = render(TransactionsList, baseProps);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the `TransactionsList` component in "compact" mode', () => {
    /** @type {import("svelte").ComponentProps<TransactionsList>} */
    const props = {
      ...baseProps,
      mode: "compact",
    };
    const { container } = render(TransactionsList, { ...baseProps, ...props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should auto-refresh relative times when the related prop is set to true", async () => {
    const props = {
      ...baseProps,
      data: {
        ...transaction,
        date: new Date(),
      },
    };
    const { container, rerender } = render(TransactionsList, props);
    const timeElement = getTimeElement(container);

    expect(timeElement.innerHTML).toMatchInlineSnapshot(`"now"`);

    await vi.advanceTimersByTimeAsync(timeRefreshInterval * 3);

    expect(timeElement.innerHTML).toMatchInlineSnapshot(`"now"`);

    await rerender({ ...props, autoRefreshTime: true });

    expect(timeElement.innerHTML).toMatchInlineSnapshot(`"3 seconds ago"`);

    await vi.advanceTimersByTimeAsync(timeRefreshInterval);

    expect(timeElement.innerHTML).toMatchInlineSnapshot(`"4 seconds ago"`);

    await vi.advanceTimersByTimeAsync(timeRefreshInterval);

    expect(timeElement.innerHTML).toMatchInlineSnapshot(`"5 seconds ago"`);
  });
});
