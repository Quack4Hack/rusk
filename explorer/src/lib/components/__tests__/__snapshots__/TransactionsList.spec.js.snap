// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transactions List > should render the \`TransactionsList\` component in "compact" mode 1`] = `
<dl
  class="detail-list"
>
  <dt
    class="details-list__term details-list__term--table"
  >
     
    ID
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <a
      class="dusk-anchor dusk-anchor--on-surface transaction-details__list-link"
      href="/some-base-path/transactions/transaction?id=4877687c2dbf154248d3ddee9ba0d81e3431f39056f82a46819da041d4ac0e04"
    >
      4877687c2dbf154248d3ddee9ba0d8...31f39056f82a46819da041d4ac0e04
    </a>
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    relative time
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <time
      class="dusk-relative-time transaction-details__list-timestamp"
      datetime="2024-04-16T06:39:09.000Z"
      slot="definition"
    >
      last month
    </time>
  </dd>
   
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Fee
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    0.000290766
     DUSK
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Status
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <span
      class="dusk-badge dusk-badge--variant-success explorer-badge"
    >
      success
    </span>
    
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Type
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <span
      class="dusk-badge dusk-badge--variant-neutral explorer-badge"
    >
      transfer
    </span>
    
  </dd>
</dl>
`;

exports[`Transactions List > should render the \`TransactionsList\` component in "full" mode 1`] = `
<dl
  class="detail-list"
>
  <dt
    class="details-list__term details-list__term--table"
  >
     
    ID
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <a
      class="dusk-anchor dusk-anchor--on-surface transaction-details__list-link"
      href="/some-base-path/transactions/transaction?id=4877687c2dbf154248d3ddee9ba0d81e3431f39056f82a46819da041d4ac0e04"
    >
      4877687c2dbf154248d3ddee9ba0d8...31f39056f82a46819da041d4ac0e04
    </a>
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    relative time
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <time
      class="dusk-relative-time transaction-details__list-timestamp"
      datetime="2024-04-16T06:39:09.000Z"
      slot="definition"
    >
      last month
    </time>
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Gas Price
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    1
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Gas Limit
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    500,000,000
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Fee
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    0.000290766
     DUSK
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Status
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <span
      class="dusk-badge dusk-badge--variant-success explorer-badge"
    >
      success
    </span>
    
  </dd>
   
  <dt
    class="details-list__term details-list__term--table"
  >
     
    Type
  </dt>
   
  <dd
    class="details-list__definition details-list__definition--table"
  >
    <span
      class="dusk-badge dusk-badge--variant-neutral explorer-badge"
    >
      transfer
    </span>
    
  </dd>
</dl>
`;
